### PR TITLE
storage: reduce scheduler msgs

### PR DESCRIPTION
**Pull request by overvenus**
## What have you changed? (mandatory)

This PR reduces scheduler msgs, also offloads some functions from scheduler to workers.

## What are the type of the changes? (mandatory)

- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

Unit tests and integration tests.

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

No.

## Does this PR affect tidb-ansible update? (mandatory)

No.

## Refer to a related PR or issue link (optional)

#3581 , #3551 and #3582

## Benchmark result if necessary (optional)

Combining these changes #3581, #3551 and #3582, QPS is improved from 40672.1 to 49144.5, and avg latnecy is reduced from 100.6ms to 83.5ms.

### CODE REVIEW

File: src/storage/txn/scheduler.rs [hunk 1]
```diff
@@ -34,129 +34,103 @@
...
-                "WritePrepareFinished [cid={}, tag={}, ctx={:?}]",
-                cid, tag, ctx
-            ),
-            Msg::WritePrepareFailed { cid, ref err } => {
-                write!(f, "WritePrepareFailed [cid={}, err={:?}]", cid, err)
-            }
-            Msg::WriteFinished { cid, .. } => write!(f, "WriteFinished [cid={}]", cid),
+            Msg::ReadFinished { ref task, .. } => write!(f, "ReadFinished [cid={}]", task.cid),
+            Msg::WriteFinished { ref task, .. } => write!(f, "WriteFinished [cid={}]", task.cid),
+            Msg::FinishedWithErr { cid, .. } => write!(f, "WriteFinished [cid={}]", cid),
```
Comment (Connor1996) [inline src/storage/txn/scheduler.rs]: wrong format message

File: src/storage/txn/process.rs [hunk 2]
```diff
@@ -619,6 +684,31 @@ impl ThreadContext for SchedContext {
     }
 }
 
+pub struct Channels<E: Engine> {
```
Comment (zhangjinpeng1987) [inline src/storage/txn/process.rs]: Maybe `Executor` is a better name?

File: src/storage/txn/process.rs [hunk 3]
```diff
@@ -565,44 +613,61 @@ fn process_write_impl<E: Engine>(
...
-        tag,
-        pr,
-        to_be_write: modifies,
-        rows,
-    }));
+    Ok((ctx, pr, modifies, rows))
+}
 
-    Ok(())
+fn handle_schedule(scheduler: worker::Scheduler<Msg>, msg: Msg) -> bool {
```
Comment (zhangjinpeng1987) [inline src/storage/txn/process.rs]: Maybe send_msg_to_scheduler is a better name. Does this function can be a member of Channels?

Comment (zhangjinpeng1987) [reply src/storage/txn/process.rs]: Or notify_scheduler?

File: src/storage/txn/process.rs [hunk 4]
```diff
@@ -147,141 +140,198 @@ impl Task {
...
                 .with_label_values(&[tag])
                 .start_coarse_timer(),
-            slow_timer: None,
         }
     }
-}
 
-impl<E: Engine> Scheduler<E> {
+    // Start the task.
+    pub fn start(&mut self) {
```
Comment (zhangjinpeng1987) [inline src/storage/txn/process.rs]: s/start/pre_start

File: src/storage/txn/process.rs [hunk 5]
```diff
@@ -205,14 +213,13 @@ impl Task {
...
+        let tag = self.tag;
         if let Some(term) = cb_ctx.term {
-            cmd.mut_context().set_term(term);
+            self.cmd.as_mut().unwrap().mut_context().set_term(term);
         }
-        let readcmd = cmd.readonly();
-        let tag = cmd.tag();
-        let pool = channels.take_pool();
-        if readcmd {
+        let pool = executor.take_pool();
```
Comment (zhangjinpeng1987) [inline src/storage/txn/process.rs]: I think it is more better to move following codes into Executor::exec, so the Executor is more like a executor.

File: src/storage/txn/process.rs [hunk 6]
```diff
@@ -147,141 +140,208 @@ impl Task {
...
+                        err: Error::from(err),
+                    },
+                );
             }
         }
     }
 
     /// Delivers a command to a worker thread for processing.
-    pub fn process_by_worker(&mut self, cid: u64, cb_ctx: CbContext, snapshot: E::Snap) {
+    fn process_by_worker<E: Engine>(
```
Comment (zhangjinpeng1987) [inline src/storage/txn/process.rs]: Maybe process_by_executor is more better.

File: src/storage/txn/process.rs [hunk 7]
```diff
@@ -147,141 +140,208 @@ impl Task {
...
-            cmd.mut_context().set_term(term);
+            self.cmd.as_mut().unwrap().mut_context().set_term(term);
         }
-        let readcmd = cmd.readonly();
-        let worker_pool = self.fetch_worker_pool(cmd.priority());
-        let tag = cmd.tag();
-        let scheduler = self.scheduler.clone();
-        if readcmd {
-            worker_pool.execute(move |ctx: &mut SchedContext| {
+        let pool = executor.take_pool();
```
Comment (zhangjinpeng1987) [inline src/storage/txn/process.rs]: It's better to wrap following codes into Executor.

File: src/storage/txn/process.rs [hunk 8]
```diff
@@ -565,44 +606,61 @@ fn process_write_impl<E: Engine>(
...
-        tag,
-        pr,
-        to_be_write: modifies,
-        rows,
-    }));
+    Ok((ctx, pr, modifies, rows))
+}
 
-    Ok(())
+fn notify_scheduler(scheduler: worker::Scheduler<Msg>, msg: Msg) -> bool {
```
Comment (Connor1996) [inline src/storage/txn/process.rs]: it can be removed in `Executor`

File: src/storage/txn/scheduler.rs [hunk 9]
```diff
@@ -210,76 +206,81 @@ impl<E: Engine> Scheduler<E> {
...
+        debug!("received new command, cid={}, cmd={}", cid, cmd);
+
+        let priority_tag = cmd.priority_tag();
+        let lock = self.gen_lock(&cmd);
+        let entity = SchedEntity::new(lock, callback, &cmd);
+        let tag = entity.tag;
+        let task = Task::new(cid, cmd);
+
+        // TODO: enqueue_task should return an reference of the entity.
+        self.enqueue_task(task, entity);
```
Comment (zhangjinpeng1987) [inline src/storage/txn/scheduler.rs]: It is weird that `enqueue_task` has a `entity` arg.

File: src/storage/txn/scheduler.rs [hunk 10]
```diff
@@ -34,129 +34,123 @@
...
+    }
 }
 
 /// Scheduler which schedules the execution of `storage::Command`s.
 pub struct Scheduler<E: Engine> {
     engine: E,
 
     // cid -> Task
-    pub cmd_ctxs: HashMap<u64, Task>,
+    pending_task: HashMap<u64, Task>,
```
Comment (Connor1996) [inline src/storage/txn/scheduler.rs]: pending_task -> pending_tasks

File: src/storage/txn/scheduler.rs [hunk 11]
```diff
@@ -450,12 +417,10 @@ impl<E: Engine> Scheduler<E> {
...
     fn acquire_lock(&mut self, cid: u64) -> bool {
-        let ctx = &mut self.cmd_ctxs.get_mut(&cid).unwrap();
-        assert_eq!(ctx.cid, cid);
-        let ok = self.latches.acquire(&mut ctx.lock, cid);
+        let entity = &mut self.schedule_entities.get_mut(&cid).unwrap();
+        let ok = self.latches.acquire(&mut entity.lock, cid);
         if ok {
-            ctx.latch_timer.take();
-            ctx.slow_timer = Some(SlowTimer::new());
+            entity.on_schedule();
```
Comment (BusyJay) [inline src/storage/txn/scheduler.rs]: Why call `on_schedule` in `acquire_lock` instead of `schedule`?

File: src/storage/txn/scheduler.rs [hunk 12]
```diff
@@ -428,14 +395,14 @@ impl<E: Engine> Scheduler<E> {
...
         if let ProcessResult::NextCommand { cmd } = pr {
             SCHED_STAGE_COUNTER_VEC
-                .with_label_values(&[ctx.tag, "next_cmd"])
+                .with_label_values(&[tag, "next_cmd"])
                 .inc();
-            self.schedule_command(cmd, cb);
+            self.schedule_command(cmd, entity.cb);
         } else {
-            execute_callback(cb, pr);
+            execute_callback(entity.cb, pr);
```
Comment (BusyJay) [inline src/storage/txn/scheduler.rs]: Why not call the callback inside thread pool?

File: src/storage/txn/scheduler.rs [hunk 13]
```diff
@@ -34,129 +34,123 @@
...
-            }
-            e @ Err(ScheduleError::Stopped(_)) => info!("scheduler worker stopped, {:?}", e),
-            Err(e) => {
-                panic!(
-                    "schedule WriteFinished msg failed, cid={}, err:{:?}",
-                    cid, e
-                );
-            }
+// TODO: Moves SchedEntity to Task once we adopt futures in the Engine trait.
+struct SchedEntity {
```
Comment (BusyJay) [inline src/storage/txn/scheduler.rs]: `Entity` doesn't seem to be a good name, it makes more sense that task itself is an entity as it contains the actual task data. How about naming it as `TaskTracker` just like `RequestTracker` in endpoint.rs?

File: src/storage/txn/scheduler.rs (hunk 12 above)
Comment (overvenus) [reply src/storage/txn/scheduler.rs]: How?

File: src/storage/txn/scheduler.rs [hunk 14]
```diff
@@ -212,28 +212,35 @@ impl<E: Engine> Scheduler<E> {
...
-        SCHED_WRITING_BYTES_GAUGE.set(self.running_write_bytes as i64);
         SCHED_CONTEX_GAUGE.set(self.pending_tasks.len() as i64);
+        if self.task_contexts.insert(cid, tctx).is_some() {
+            panic!("TaskContext cid={} shouldn't exist", cid);
+        }
     }
 
-    fn dequeue_schedule_entity(&mut self, cid: u64) -> SchedEntity {
-        let entity = self.schedule_entities.remove(&cid).unwrap();
+    fn find_task_context(&mut self, cid: u64) -> TaskContext {
```
Comment (zhangjinpeng1987) [inline src/storage/txn/scheduler.rs]: s/find_task_context/take_task_context

**Pull request merged**